In [ ]:
20

In [ ]:
20

In [2]:
from federatedscope.register import register_data
from federatedscope.register import register_trainer
from federatedscope.register import register_metric
from federatedscope.register import register_model




### Register model

In [3]:
#from federatedscope.contrib.model.mnist_model import call_my_net
#register_model("mynet", call_my_net)

### Register trainer

In [4]:


from federatedscope.contrib.trainer.laplacian_trainer_with_domain_separation_with_summation_1MINE_VAE_separated import call_laplacian_trainer

register_trainer('laplacian_trainer', call_laplacian_trainer)

### Register metric

### Create the config

#### Set data, model, trainer and metric

In [5]:
from federatedscope.core.configs.config import global_cfg, CN
cfg = global_cfg.clone()

cfg.merge_from_file("scripts/B-FHTL_exp_scripts/Graph-DT/fed_dom_sep_MI.yaml")
cfg.data.save_dir = 'test_dir'
# cfg.data.type = 'cikm_cup'
#cfg.data.root = 'data'
#cfg.data.shuffle=True
#cfg.data.transform = [['ToTensor'], ['Normalize', {'mean': [0.], 'std': [1]}]]
#cfg.model.type = 'gin'
#cfg.model.out_channels = 10
#cfg.model.hidden = 64
#cfg.model.task='graph'
#cfg.model.dropout = 0.5
#cfg.personalization.local_param = ['encoder_atom', 'encoder', 'clf']#['node_encoder', 'clf']
#cfg.train.batch_or_epoch = "epoch"
cfg.trainer.type = 'laplacian_trainer'
cfg.data.batch_size = 64
# cfg.eval.metric = ['mymetric']
cfg.params = CN()
cfg.params.alpha=0.1
cfg.params.csd_importance= 1e6
cfg.params.diff_importance = 0.1  # 1
cfg.params.mine_lr = 0.01
cfg.params.eps=1e-20
cfg.params.p=0.
cfg.params.lam = 0.
cfg.params.recon_importance = 0.1
cfg.params.kld_importance = 0.1


#### configure other options

In [6]:
#cfg.use_gpu = True 0.05 [0.02-0.03] 0.04-0.06
#cfg.best_res_update_round_wise_key = "test_loss"

#cfg.federate.mode = 'standalone'

cfg.federate.method = \
    f'Laplacian_MINE_VAE_test_no_KLD'

#cfg.federate.local_update_steps = 20000000
cfg.personalization.local_update_steps = 20000000
#cfg.finetune.local_update_steps = 20000000
#cfg.train.local_update_steps = 1

cfg.federate.total_round_num = 2000
cfg.federate.client_num = 2
cfg.early_stop.patience = 3000
#cfg.train.optimizer.lr = 0.001
#cfg.train.optimizer.weight_decay = 0.0005
#cfg.grad.grad_clip = 2.0
cfg.criterion.type = 'CrossEntropyLoss'
#cfg.seed = 123
cfg.eval.freq = 1
cfg.eval.metrics = ['imp_ratio']
cfg.eval.report = ['avg']
cfg.eval.best_res_update_round_wise_key = 'val_imp_ratio'


In [7]:
import torch
#torch.manual_seed(0)
#torch.backends.cudnn.benchmark = False
#torch.use_deterministic_algorithms(F/home/michael/Desktop/backup files/FederatedScope/data/CIKM22Competitionalse)
#import random
#random.seed(0)
#import numpy as np
#np.random.seed(0)

In [8]:
from yacs.config import CfgNode
client_cfg_file = "scripts/B-FHTL_exp_scripts/Graph-DT/cfg_per_client_theirs.yaml"# "scripts/B-FHTL_exp_scripts/Graph-DT/cfg_per_client_ours_lr_local_steps.yaml"
client_cfg = CfgNode.load_cfg(open(client_cfg_file,
                                       'r')) if client_cfg_file else None


### Start the FL prosess

In [9]:
from federatedscope.core.auxiliaries.data_builder import get_data
from federatedscope.core.auxiliaries.utils import setup_seed, update_logger
from federatedscope.core.fed_runner import FedRunner
from federatedscope.core.auxiliaries.worker_builder import get_server_cls, get_client_cls


In [10]:
from federatedscope.contrib.workers.laplacian_with_domain_separation_MI_client import LaplacianDomainSeparationMIClient
from federatedscope.contrib.workers.laplacian_server_dom_sep import LaplacianServerDomSep

setup_seed(cfg.seed)
update_logger(cfg)
data, modified_cfg = get_data(cfg)
cfg.merge_from_other_cfg(modified_cfg)


2023-01-12 15:03:35,283 (utils:129)INFO: the current machine is at 127.0.1.1
2023-01-12 15:03:35,284 (utils:131)INFO: the current dir is /home/michael/Master-Thesis/CKIM_Competition
2023-01-12 15:03:35,285 (utils:132)INFO: the output dir is exp/test_dir/Laplacian_MINE_VAE_test_no_KLD_gin_on_graph-dt_lr0.1_lstep1_/sub_exp_20230112150335


In [12]:
from federatedscope.contrib.workers.laplacian_with_domain_separation_1MINE_VAE_Separated_client import LaplacianDomainSeparation1MINE_Separated_Client
from federatedscope.contrib.workers.laplacian_server import LaplacianServer

Fed_runner = FedRunner(data=data,
                       server_class=LaplacianServer,
                       client_class=LaplacianDomainSeparation1MINE_Separated_Client,
                       config=cfg.clone(),
                       client_config=client_cfg)
Fed_runner.run()

/home/michael/Master-Thesis/CKIM_Competition/venv/lib/python3.9/site-packages/torch/nn/init.py:405: UserWarning: Initializing zero-element tensors is a no-op
  warnings.warn("Initializing zero-element tensors is a no-op")
2023-01-12 15:07:45,551 (fed_runner:249)INFO: Server #0 has been set up ... 


rho: 0.0
server params: 
encoder_atom.atom_embedding_list.0.weight
encoder_atom.atom_embedding_list.1.weight
encoder_atom.atom_embedding_list.2.weight
encoder_atom.atom_embedding_list.3.weight
encoder_atom.atom_embedding_list.4.weight
encoder_atom.atom_embedding_list.5.weight
encoder_atom.atom_embedding_list.6.weight
encoder_atom.atom_embedding_list.7.weight
encoder_atom.atom_embedding_list.8.weight
encoder_atom.atom_embedding_list.9.weight
encoder_atom.atom_embedding_list.10.weight
encoder_atom.atom_embedding_list.11.weight
encoder_atom.atom_embedding_list.12.weight
encoder_atom.atom_embedding_list.13.weight
encoder_atom.atom_embedding_list.14.weight
encoder_atom.atom_embedding_list.15.weight
encoder_atom.atom_embedding_list.16.weight
encoder_atom.atom_embedding_list.17.weight
encoder_atom.atom_embedding_list.18.weight
encoder_atom.atom_embedding_list.19.weight
encoder_atom.atom_embedding_list.20.weight
encoder_atom.atom_embedding_list.21.weight
encoder.weight
encoder.bias
local_gnn.c

2023-01-12 15:07:45,606 (config:261)INFO: the used configs are: 
asyn:
  min_received_num: 2
  min_received_rate: -1.0
  timeout: 0
  use: True
attack:
  alpha_TV: 0.001
  alpha_prop_loss: 0
  attack_method: 
  attacker_id: -1
  classifier_PIA: randomforest
  info_diff_type: l2
  inject_round: 0
  max_ite: 400
  reconstruct_lr: 0.01
  reconstruct_optim: Adam
  target_label_ind: -1
backend: torch
cfg_file: 
criterion:
  type: MSELoss
data:
  args: []
  batch_size: 64
  cSBM_phi: [0.5, 0.5, 0.5]
  consistent_label_distribution: False
  drop_last: False
  graphsaint:
    num_steps: 30
    walk_length: 2
  loader: 
  num_workers: 0
  pre_transform: []
  quadratic:
    dim: 1
    max_curv: 12.5
    min_curv: 0.02
  root: data/
  save_dir: test_dir
  server_holds_all: False
  shuffle: True
  sizes: [10, 5]
  splits: [0.8, 0.1, 0.1]
  splitter: ooxx
  splitter_args: []
  subsample: 1.0
  target_transform: []
  transform: []
  type: graph-dt
device: 0
distribute:
  use: False
early_stop:
  del

rho: 0.0


2023-01-12 15:07:45,833 (config:261)INFO: the used configs are: 
asyn:
  min_received_num: 2
  min_received_rate: -1.0
  timeout: 0
  use: True
attack:
  alpha_TV: 0.001
  alpha_prop_loss: 0
  attack_method: 
  attacker_id: -1
  classifier_PIA: randomforest
  info_diff_type: l2
  inject_round: 0
  max_ite: 400
  reconstruct_lr: 0.01
  reconstruct_optim: Adam
  target_label_ind: -1
backend: torch
cfg_file: 
criterion:
  type: MSELoss
data:
  args: []
  batch_size: 64
  cSBM_phi: [0.5, 0.5, 0.5]
  consistent_label_distribution: False
  drop_last: False
  graphsaint:
    num_steps: 30
    walk_length: 2
  loader: 
  num_workers: 0
  pre_transform: []
  quadratic:
    dim: 1
    max_curv: 12.5
    min_curv: 0.02
  root: data/
  save_dir: test_dir
  server_holds_all: False
  shuffle: True
  sizes: [10, 5]
  splits: [0.8, 0.1, 0.1]
  splitter: ooxx
  splitter_args: []
  subsample: 1.0
  target_transform: []
  transform: []
  type: graph-dt
device: 0
distribute:
  use: False
early_stop:
  del

rho: 0.0
in train
mi_local_global: 0.00022700056433677673
mi_global_fixed: 0.0015235953032970428
rec_loss: 1.0000710487365723
kld_loss: 1.7968862056732178
mi_local_global: 0.0009042136371135712
mi_global_fixed: 0.00020004436373710632
rec_loss: 0.9880773425102234
kld_loss: 1.802543044090271
mi_local_global: 0.0006496645510196686
mi_global_fixed: 0.0019105393439531326
rec_loss: 0.9965176582336426
kld_loss: 1.7663557529449463
mi_local_global: 0.0005304440855979919
mi_global_fixed: 0.0017413273453712463
rec_loss: 1.0013744831085205
kld_loss: 1.7783476114273071
mi_local_global: 0.0023493096232414246
mi_global_fixed: 0.0013433285057544708
rec_loss: 0.9998010993003845
kld_loss: 1.7787655591964722
mi_local_global: 0.0004069991409778595
mi_global_fixed: 0.0012850947678089142
rec_loss: 0.9885938763618469
kld_loss: 1.8024003505706787
mi_local_global: 0.0015517808496952057
mi_global_fixed: 0.00042789801955223083
rec_loss: 1.0076079368591309
kld_loss: 1.7833516597747803
mi_local_global: -0.00095918

2023-01-12 15:08:01,175 (laplacian_with_domain_separation_1MINE_VAE_Separated_client:144)INFO: {'Role': 'Client #1', 'Round': 0, 'Results_raw': {'train_imp_ratio': -977.98225, 'train_total': 901, 'train_loss': 8977.374865, 'train_avg_loss': 9.96379}}


in train
mi_local_global: 0.0011111460626125336
mi_global_fixed: 0.0032234564423561096
rec_loss: 1.010201334953308
kld_loss: 1.7939069271087646
mi_local_global: -0.0002789013087749481
mi_global_fixed: -0.0002977512776851654
rec_loss: 1.0116615295410156
kld_loss: 1.8166829347610474
mi_local_global: -0.00360952690243721
mi_global_fixed: -0.0058601014316082
rec_loss: 0.9933339357376099
kld_loss: 1.7620728015899658
mi_local_global: 4.993751645088196e-05
mi_global_fixed: -0.006067398935556412
rec_loss: 0.9905187487602234
kld_loss: 1.7900359630584717
mi_local_global: -0.005377002060413361
mi_global_fixed: -0.006818361580371857
rec_loss: 0.9931280016899109
kld_loss: 1.7589975595474243
mi_local_global: -0.004402082413434982
mi_global_fixed: -0.004358615726232529
rec_loss: 0.9798267483711243
kld_loss: 1.7386164665222168
mi_local_global: -0.00077071413397789
mi_global_fixed: -0.00488324835896492
rec_loss: 0.9855772852897644
kld_loss: 1.7401349544525146
mi_local_global: -0.0003199055790901184
mi_

2023-01-12 15:08:06,863 (laplacian_with_domain_separation_1MINE_VAE_Separated_client:144)INFO: {'Role': 'Client #2', 'Round': 0, 'Results_raw': {'train_imp_ratio': -4169.809093, 'train_total': 512, 'train_loss': 579.327694, 'train_avg_loss': 1.131499}}
2023-01-12 15:08:07,060 (laplacian_server:160)INFO: Server #0: Starting evaluation at the end of round 0.
2023-01-12 15:08:07,081 (laplacian_server:167)INFO: ----------- Starting a new training round (Round #1) -------------


in val or test
mi_local_global: -0.0008509568870067596
mi_global_fixed: -0.0006540119647979736
rec_loss: 1.0611292123794556
kld_loss: 1.773118495941162
mi_local_global: -0.0005799755454063416
mi_global_fixed: -0.0003493502736091614
rec_loss: 1.0564842224121094
kld_loss: 1.827627182006836
in val or test
mi_local_global: -0.0005360059440135956
mi_global_fixed: -0.00045745447278022766
rec_loss: 1.0609939098358154
kld_loss: 1.809572696685791


2023-01-12 15:08:08,075 (client:410)INFO: {'Role': 'Client #1', 'Round': 1, 'Results_raw': {'test_imp_ratio': -722.37633, 'test_total': 113, 'test_loss': 858.938336, 'test_avg_loss': 7.601224, 'val_imp_ratio': -710.831743, 'val_total': 113, 'val_loss': 846.880453, 'val_avg_loss': 7.494517}}
2023-01-12 15:08:08,076 (monitor:513)INFO: current_best=-710.831743, should_save=True
2023-01-12 15:08:08,078 (client:431)INFO: Client: #1, val_imp_ratio: -710.831743. model saved at exp/test_dir/Laplacian_MINE_VAE_test_no_KLD_gin_on_graph-dt_lr0.1_lstep1_/sub_exp_20230112150335/model1.pth


mi_local_global: -0.0005084015429019928
mi_global_fixed: -0.00043707340955734253
rec_loss: 1.0625025033950806
kld_loss: 1.801452875137329
in val or test
mi_local_global: -0.00030851736664772034
mi_global_fixed: -0.0004592984914779663
rec_loss: 1.0165774822235107
kld_loss: 1.768758773803711
in val or test
mi_local_global: -0.0002649165689945221
mi_global_fixed: -0.00037946924567222595
rec_loss: 1.0110129117965698
kld_loss: 1.7267193794250488


2023-01-12 15:08:08,688 (client:410)INFO: {'Role': 'Client #2', 'Round': 1, 'Results_raw': {'test_imp_ratio': -1314.202632, 'test_total': 64, 'test_loss': 23.984875, 'test_avg_loss': 0.374764, 'val_imp_ratio': -1424.08132, 'val_total': 63, 'val_loss': 25.444541, 'val_avg_loss': 0.403882}}
2023-01-12 15:08:08,689 (monitor:513)INFO: current_best=-1424.08132, should_save=True
2023-01-12 15:08:08,691 (client:431)INFO: Client: #2, val_imp_ratio: -1424.08132. model saved at exp/test_dir/Laplacian_MINE_VAE_test_no_KLD_gin_on_graph-dt_lr0.1_lstep1_/sub_exp_20230112150335/model2.pth


mi_local_global: 0.0002762824296951294
mi_global_fixed: 0.0003419443964958191
rec_loss: 1.007103443145752
kld_loss: 1.8186930418014526
mi_local_global: 0.0012037530541419983
mi_global_fixed: 0.0005493797361850739
rec_loss: 1.0125761032104492
kld_loss: 1.7638535499572754
mi_local_global: -0.0003543887287378311
mi_global_fixed: 0.002388596534729004
rec_loss: 1.0074539184570312
kld_loss: 1.7693732976913452
mi_local_global: 0.0007153935730457306
mi_global_fixed: -0.0004081428050994873
rec_loss: 1.0202220678329468
kld_loss: 1.7778782844543457
mi_local_global: -0.0006461441516876221
mi_global_fixed: 0.0004102736711502075
rec_loss: 1.0069235563278198
kld_loss: 1.807395339012146
mi_local_global: 0.0028726235032081604
mi_global_fixed: 0.001431906595826149
rec_loss: 1.0027128458023071
kld_loss: 1.7855173349380493
mi_local_global: 0.0003602616488933563
mi_global_fixed: 0.001367766410112381
rec_loss: 1.0075322389602661
kld_loss: 1.7995431423187256
mi_local_global: -0.0009671598672866821
mi_global_

2023-01-12 15:08:19,156 (laplacian_with_domain_separation_1MINE_VAE_Separated_client:144)INFO: {'Role': 'Client #1', 'Round': 1, 'Results_raw': {'train_imp_ratio': -617.331576, 'train_total': 901, 'train_loss': 5973.896621, 'train_avg_loss': 6.630296}}


mi_local_global: 0.002810399979352951
mi_global_fixed: -0.00710558146238327
rec_loss: 0.9855462908744812
kld_loss: 1.7527782917022705
mi_local_global: 0.003614259883761406
mi_global_fixed: -0.006189070641994476
rec_loss: 0.9792604446411133
kld_loss: 1.7598791122436523
mi_local_global: 0.0035458803176879883
mi_global_fixed: -0.005202349275350571
rec_loss: 0.9834400415420532
kld_loss: 1.7542996406555176
mi_local_global: 0.0022410228848457336
mi_global_fixed: -0.006157424300909042
rec_loss: 0.9855536818504333
kld_loss: 1.7160720825195312
mi_local_global: 0.001897074282169342
mi_global_fixed: -0.005397595465183258
rec_loss: 0.9840453267097473
kld_loss: 1.7559174299240112
mi_local_global: 0.002655845135450363
mi_global_fixed: -0.004743408411741257
rec_loss: 0.9883403778076172
kld_loss: 1.7142722606658936
mi_local_global: 0.001379169523715973
mi_global_fixed: -0.0034188590943813324
rec_loss: 0.985285758972168
kld_loss: 1.740079641342163
mi_local_global: 0.006020167842507362
mi_global_fixed: 

2023-01-12 15:08:24,746 (laplacian_with_domain_separation_1MINE_VAE_Separated_client:144)INFO: {'Role': 'Client #2', 'Round': 1, 'Results_raw': {'train_imp_ratio': -1193.186872, 'train_total': 512, 'train_loss': 175.459586, 'train_avg_loss': 0.342695}}
2023-01-12 15:08:24,756 (server:480)INFO: {'Role': 'Server #', 'Round': 1, 'Results_avg': {'test_imp_ratio': -1018.289481, 'test_total': 88.5, 'test_loss': 441.461605, 'test_avg_loss': 3.987994, 'val_imp_ratio': -1067.456532, 'val_total': 88.0, 'val_loss': 436.162497, 'val_avg_loss': 3.949199}}
2023-01-12 15:08:24,757 (monitor:513)INFO: current_best=-10000, should_save=False
2023-01-12 15:08:24,758 (monitor:513)INFO: current_best=-1067.456532, should_save=True
2023-01-12 15:08:24,954 (laplacian_server:160)INFO: Server #0: Starting evaluation at the end of round 1.
2023-01-12 15:08:24,978 (laplacian_server:167)INFO: ----------- Starting a new training round (Round #2) -------------


in val or test
mi_local_global: -0.0008099228143692017
mi_global_fixed: -0.0006941333413124084
rec_loss: 1.0809502601623535
kld_loss: 1.7728452682495117
mi_local_global: 0.00012383237481117249
mi_global_fixed: 0.0006363727152347565
rec_loss: 1.079336166381836
kld_loss: 1.8272775411605835
in val or test
mi_local_global: -5.089864134788513e-05
mi_global_fixed: -0.00016207247972488403
rec_loss: 1.0841923952102661
kld_loss: 1.8092641830444336


2023-01-12 15:08:25,903 (client:410)INFO: {'Role': 'Client #1', 'Round': 2, 'Results_raw': {'test_imp_ratio': -614.102775, 'test_total': 113, 'test_loss': 745.851069, 'test_avg_loss': 6.600452, 'val_imp_ratio': -594.91667, 'val_total': 113, 'val_loss': 725.811913, 'val_avg_loss': 6.423114}}
2023-01-12 15:08:25,904 (monitor:513)INFO: current_best=-594.91667, should_save=True
2023-01-12 15:08:25,906 (client:431)INFO: Client: #1, val_imp_ratio: -594.91667. model saved at exp/test_dir/Laplacian_MINE_VAE_test_no_KLD_gin_on_graph-dt_lr0.1_lstep1_/sub_exp_20230112150335/model1.pth


mi_local_global: 0.0005807802081108093
mi_global_fixed: 0.0007863417267799377
rec_loss: 1.0853604078292847
kld_loss: 1.8011512756347656
in val or test
mi_local_global: 0.0005939975380897522
mi_global_fixed: -0.0006612688302993774
rec_loss: 1.0529330968856812
kld_loss: 1.7670819759368896
in val or test
mi_local_global: 0.0007601790130138397
mi_global_fixed: -0.00041875988245010376
rec_loss: 1.054380178451538
kld_loss: 1.7254512310028076


2023-01-12 15:08:26,524 (client:410)INFO: {'Role': 'Client #2', 'Round': 2, 'Results_raw': {'test_imp_ratio': -572.632793, 'test_total': 64, 'test_loss': 11.407852, 'test_avg_loss': 0.178248, 'val_imp_ratio': -795.051911, 'val_total': 63, 'val_loss': 14.942894, 'val_avg_loss': 0.237189}}
2023-01-12 15:08:26,525 (monitor:513)INFO: current_best=-795.051911, should_save=True
2023-01-12 15:08:26,526 (client:431)INFO: Client: #2, val_imp_ratio: -795.051911. model saved at exp/test_dir/Laplacian_MINE_VAE_test_no_KLD_gin_on_graph-dt_lr0.1_lstep1_/sub_exp_20230112150335/model2.pth


in train
mi_local_global: 0.004123980179429054
mi_global_fixed: -0.003421604633331299
rec_loss: 0.9835219383239746
kld_loss: 1.7261648178100586
mi_local_global: 0.0008323825895786285
mi_global_fixed: -0.004303723573684692
rec_loss: 0.9898428916931152
kld_loss: 1.7536691427230835
mi_local_global: 0.0032435618340969086
mi_global_fixed: -0.003322012722492218
rec_loss: 0.9836898446083069
kld_loss: 1.7106976509094238


KeyboardInterrupt: 

client 0 times: 0.346s
bei normal fedavg: 0.15s
mit MINE: 1.7 Sekunden


+

In [ ]:
777-431


In [ ]:
data

In [ ]:
55: -41 with 0 KLD
mit 0.1 KLD: -31

In [ ]:
import os

,,0After 10 rounds:

diff: 0, csd: 0:
server: -8
c1 t/val: 0.375 /0.37
c2 t/val: 0.678 /0.606
c3 t/val: 0.699 /0.608
c4 t/val: 0.599 /0.537

diff: 0, csd: 0:
server: -4.5
c1 t/val: 0.372879 /0.328
c2 t/val: 0.678992 /0.605809
c3 t/val: 0.70166 /0.608609
c4 t/val: 0.597455 /0.534253



diff: 10, csd: 0:
server: -35
c1 t/val: 0.482/0.52
c2 t/val: 0.694/0.65
c3 t/val: 0.724/0.65
c4 t/val: 0.674/0.575

diff: 1, csd: 0:
server: -9.6
c1 t/val: 0.378194/0.319349
c2 t/val: 0.67461/0.609163
c3 t/val: 0.705559/0.615019
c4 t/val: 0.605581/0.572086

diff: 0, csd: 1e4:
server: -18
c1 t/val: 0.280962/0.23074
c2 t/val: 0.682652/0.603175
c3 t/val: 0.699094/0.67027
c4 t/val: 0.594985/0.609611

diff: 0 csd: 1e8
server: -18
c1 t/val: 0.2606/0.348(0.197)
c2 t/val: 0.653/0.0.602
c3 t/val: 0.691/0.597
c4 t/val: 0.581/0.629

diff: 0 csd: 1e8
server: -12.7
c1 t/val: 0.262647/0.249875
c2 t/val: 0.649273/0.0.600435
c3 t/val: 0.696132/0.615158
c4 t/val: 0.586597/0.656992


diff: 0.1 csd: 0
server: -11
c1 t/val: 0.2606/0.332837
c2 t/val: 0.678326/0.0.610099
c3 t/val: 0.704083/0.616179
c4 t/val: 0.604177/0.559658




In [ ]:
os.environ['CUBLAS_WORKSPACE_CONFIG']